crates/ruma-server-util/src/authorization.rs

```diff
@@ -37,204 +43,164 @@ impl XMatrix {
+impl fmt::Debug for XMatrix {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("XMatrix")
+            .field("origin", &self.origin)
+            .field("destination", &self.destination)
+            .field("key", &self.key)
+            .finish_non_exhaustive()
+    }
 }
 
-fn is_digit(c: u8) -> bool {
-    (0x30..=0x39).contains(&c)
-}
+impl fmt::Display for XMatrix {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self { origin, destination, key, sig } = self;
 
-fn is_tchar(c: u8) -> bool {
-    const TOKEN_CHARS: [u8; 15] =
-        [b'!', b'#', b'$', b'%', b'&', b'\'', b'*', b'+', b'-', b'.', b'^', b'_', b'`', b'|', b'~'];
-    is_alpha(c) || is_digit(c) || TOKEN_CHARS.contains(&c)
-}
+        write!(f, r#"X-Matrix origin="{origin}""#)?;
+
+        if let Some(destination) = destination {
+            write!(f, r#",destination="{destination}""#)?;
+        }
 
-fn is_qdtext(c: u8) -> bool {
-    c == b'\t'
-        || c == b' '
-        || c == 0x21
-        || (0x23..=0x5B).contains(&c)
-        || (0x5D..=0x7E).contains(&c)
-        || is_obs_text(c)
+        write!(f, r#",key="{key}",sig="{sig}""#)
+    }
 }
 
-fn is_obs_text(c: u8) -> bool {
-    c >= 0x80 // The spec does contain an upper limit of 0xFF here, but that's enforced by the type
+impl FromStr for XMatrix {
+    type Err = XMatrixParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::parse(s)
+    }
 }
 
-fn is_vchar(c: u8) -> bool {
-    (0x21..=0x7E).contains(&c)
+impl TryFrom<&HeaderValue> for XMatrix {
+    type Error = XMatrixParseError;
+
+    fn try_from(value: &HeaderValue) -> Result<Self, Self::Error> {
+        Self::parse(value.to_str()?)
+    }
 }
 
-fn is_quoted_pair(c: u8) -> bool {
-    c == b'\t' || c == b' ' || is_vchar(c) || is_obs_text(c)
+impl TryFrom<&XMatrix> for HeaderValue {
+    type Error = http::header::InvalidHeaderValue;
+
+    fn try_from(value: &XMatrix) -> Result<Self, Self::Error> {
+        value.to_string().try_into()
+    }
 }
 
 impl Credentials for XMatrix {
     const SCHEME: &'static str = "X-Matrix";
 
     fn decode(value: &HeaderValue) -> Option<Self> {
-        let value: Vec<u8> = value.as_bytes().to_vec();
-        parse_xmatrix(&mut value.into_tokens())
+        value.try_into().ok()
     }
 
     fn encode(&self) -> HeaderValue {
-        if let Some(destination) = &self.destination {
-            format!(
-                "X-Matrix origin=\"{}\",destination=\"{destination}\",key=\"{}\",sig=\"{}\"",
-                self.origin, self.key, self.sig
-            )
-        } else {
-            format!("X-Matrix origin=\"{}\",key=\"{}\",sig=\"{}\"", self.origin, self.key, self.sig)
-        }
-        .try_into()
-        .expect("header format is static")
+        self.try_into().expect("header format is static")
+    }
+}
+
+/// An error when trying to parse a X-Matrix Authorization header.
```

Comment:
"X" is pronounced starting with a vowel (e).
```suggestion
/// An error when trying to parse an X-Matrix Authorization header.
```

Reply:
Fixed.